scripts/build-toolchains.sh

```diff
@@ -8,31 +10,149 @@ unamestr=$(uname)
+    echo "   hwacha: if set, builds esp-tools toolchain"
+    echo "   ec2fast: if set, pulls in a pre-compiled RISC-V toolchain for an EC2 manager instance"
+}
+
+#taken from riscv-tools to check for open-ocd autoconf versions
+check_version() {
+    $1 --version | awk "NR==1 {if (\$NF>$2) {exit 0} exit 1}" || (
+        echo $3 requires at least version $2 of $1. Aborting.
+        exit 1
+    )
+}
+
+if [ "$1" == "--help" -o "$1" == "-h" -o "$1" == "-H" ]; then
+    usage
+    exit 3
+fi
+
+TOOLCHAIN="riscv-tools"
+EC2FASTINSTALL="false"
+FASTINSTALL="false"
+while test $# -gt 0
+do
+   case "$1" in
+        riscv)
+            TOOLCHAIN="riscv-tools"
+            ;;
+        hwacha)
+            TOOLCHAIN="esp-tools"
+            ;;
+        ec2fast | --ec2fast) # I don't want to break this api
+            EC2FASTINSTALL=true
+            ;;
+        -h | -H | --help)
+            usage
+            exit
+            ;;
+        --*) echo "ERROR: bad option $1"
+            usage
+            exit 1
+            ;;
+        *) echo "ERROR: bad argument $1"
+            usage
+            exit 2
+            ;;
+    esac
+    shift
+done
+
+
+if [ "$EC2FASTINSTALL" = "true" ]; then
+    if [ "$TOOLCHAIN" = "riscv-tools" ]; then
+      cd $RDIR
+      git clone https://github.com/firesim/firesim-riscv-tools-prebuilt.git
+      cd firesim-riscv-tools-prebuilt
+      git checkout $PRECOMPILED_REPO_HASH
+      PREBUILTHASH="$(cat HASH)"
+      git -C $CHIPYARD_DIR submodule update --init  toolchains/$TOOLCHAIN
+      cd "$CHIPYARD_DIR/toolchains/$TOOLCHAIN"
+      GITHASH="$(git rev-parse HEAD)"
+      cd $RDIR
+      echo "prebuilt hash: $PREBUILTHASH"
+      echo "git      hash: $GITHASH"
+      if [[ $PREBUILTHASH == $GITHASH && "$EC2FASTINSTALL" == "true" ]]; then
+          FASTINSTALL=true
+          echo "Using fast pre-compiled install for riscv-tools"
+      else
+          echo "Error: hash of precompiled toolchain doesn't match the riscv-tools submodule hash."
+          exit
+      fi
+    else
+          echo "Error: No precompiled toolchain for esp-tools or other non-native riscv-tools."
+          exit 
+    fi
 fi
 
 INSTALL_DIR="$TOOLCHAIN-install"
-mkdir -p "$(pwd)/$INSTALL_DIR"
 
 RISCV="$(pwd)/$INSTALL_DIR"
 
 # install risc-v tools
 export RISCV="$RISCV"
-git -C $CHIPYARD_DIR submodule update --init --recursive toolchains/$TOOLCHAIN #--jobs 8
-cd "$CHIPYARD_DIR/toolchains/$TOOLCHAIN"
-export MAKEFLAGS="-j16"
-./build.sh
+
+if [ "$FASTINSTALL" = true ]; then
+    cd firesim-riscv-tools-prebuilt
+    ./installrelease.sh
+    mv distrib "$RISCV"
+    # copy HASH in case user wants it later
+    cp HASH "$RISCV"
+    cd $RDIR
+    rm -rf firesim-riscv-tools-prebuilt
+else
+    mkdir -p "$RISCV"
+    git -C $CHIPYARD_DIR submodule update --init --recursive toolchains/$TOOLCHAIN #--jobs 8
+    cd "$CHIPYARD_DIR/toolchains/$TOOLCHAIN"
+    export MAKEFLAGS="-j16"
+    #build the actual toolchain
+    #./build.sh
```

Comment:
We inlined this to avoid building openocd?

Reply:
Yes, and since we need to build the linux version of the toolchain as well anyways